libsweep/src/sweep.cc

```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
   uint8_t args[2] = {0};
   sweep::protocol::integral_to_ascii_bytes(code, args);
 
-  sweep::protocol::error_s protocolerror = nullptr;
+  sweep::protocol::write_command_with_arguments(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, args);
 
-  sweep::protocol::write_command_with_arguments(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, args, &protocolerror);
+  sweep::protocol::response_param_s response;
+  sweep::protocol::read_response_param(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, &response);
+}
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to send sample rate command");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
-  }
+void sweep_device_reset(sweep_device_s device) {
+  SWEEP_ASSERT(device);
+  SWEEP_ASSERT(!device->is_scanning);
 
-  sweep::protocol::response_param_s response;
-  sweep::protocol::read_response_param(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, &response, &protocolerror);
+  sweep::protocol::write_command(device->serial, sweep::protocol::RESET_DEVICE);
+}
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to receive sample rate command response");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
+template <class F> auto translateException(F&& f, sweep_error_s* error) -> decltype(std::forward<F>(f)()) {
+  try {
+    return std::forward<F>(f)();
+  } catch (const sweep::ErrorBase& e) {
+    *error = new sweep_error{e.what()};
+  } catch (const std::bad_alloc&) {
+    *error = new sweep_error{"Could not allocate enough memory"};
+  } catch (...) {
+    *error = new sweep_error{"Unknown exception"};
+  }
+  return {};
+}
+
+template <class F> void translateExceptionV(F&& f, sweep_error_s* error) {
+  try {
+    std::forward<F>(f)();
+  } catch (const sweep::ErrorBase& e) {
+    *error = new sweep_error{e.what()};
+  } catch (const std::bad_alloc&) {
+    *error = new sweep_error{"Could not allocate enough memory"};
+  } catch (...) {
+    *error = new sweep_error{"Unknown exception"};
   }
 }
+} // local namespace
+
+/* ###################### implementation of public API ################################# */
```

Comment:
Separation of concerns: The private functions are concerned with implementing the functionality the implementation of the public functions then only needs to take care about eror translation and any other adaptions to the ABI that might be necessary in the future. Also `translateException` needs some function / function object to be passed to it anyway, so the only alternative I see would be to wrapp the internal functionality into a lambda which doesn't really make the code more readable.